### PR TITLE
Add lifecycle status check of API to Authentication.bal

### PR DESCRIPTION
**Pull request by dewmini**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/constants/Constants.bal [hunk 1]
```diff
@@ -5,7 +5,7 @@ const string GW_HOME = "GW_HOME";
 const string API_CREATE = "API_CREATE";
 const string API_UPDATE = "API_UPDATE";
 const string API_DELETE = "API_DELETE";
-const string API_PUBLISH = "API_PUBLISH";
+const string API_STATE_CHANGE = "API_STATE_CHANGE";fmain
```
Comment (pubudu538) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/constants/Constants.bal]: fmain?

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/apimPublisherEventListner.bal [hunk 2]
```diff
@@ -74,8 +74,16 @@ service jmsService {
...
-                //TODO: Implement logic
+            } else if(strings:equalsIgnoreCase(eventType, Constants:API_STATE_CHANGE)){
+                json apiSummary = jsons:getJson(event, "apiSummary");
+                if(apiSummary != null){
+
+                //Update API cache
+                dto:APIDTO api = gatewayUtil:fromJSONToAPIDTO(apiSummary);
+                holder:putIntoAPICache(api);
+                } else {
+                system:println("Invalid json received");
```
Comment (isharac) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/apimPublisherEventListner.bal]: indentation?

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/utils/Utils.bal [hunk 3]
```diff
@@ -20,6 +20,10 @@ function constructSubscriptionNotFound (message response) {
     json payload = {"code":900903, "message":"subscription not found"};
     messages:setJsonPayload(response, payload);
 }
+function constructAPIIsInMaintenance (message response) {
```
Comment (praminda) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/utils/Utils.bal]: File is not formatted properly. There should be a new line between methods

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/utils/Utils.bal [hunk 4]
```diff
@@ -20,6 +20,10 @@ function constructSubscriptionNotFound (message response) {
...
+    json payload = {"code":700700, "message":"This API has been blocked temporarily"};
+    messages:setJsonPayload(response, payload);
+}
```
Comment (praminda) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/utils/Utils.bal]: Same here

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/Authentication.bal [hunk 5]
```diff
@@ -40,6 +40,17 @@ function authenticate (message m) (boolean, message) {
     string version = "1.0.0";
     string uriTemplate = "/*";
     string httpVerb = strings:toUpperCase(http:getMethod(m));
+    //check api status
+    string apiKey = apiContext + ":" + version;
+    dto:APIDTO apiDto = holder:getFromAPICache(apiKey);
+
+    if (constants:MAINTENANCE == apiDto.lifeCycleStatus) {
```
Comment (sambaheerathan) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/Authentication.bal]: Don't we need to do a null check for apiDto before using it ?

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/constants/Constants.bal (hunk 1 above)
Comment (dewmini) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/constants/Constants.bal]: Fixed in https://github.com/wso2/carbon-apimgt/commit/f089d81b6be2a0e8ddf8e530317fe680abafd246

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/Authentication.bal (hunk 5 above)
Comment (dewmini) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/Authentication.bal]: Fixed in https://github.com/wso2/carbon-apimgt/pull/4087/commits/e9aae8173ce0104f16c45795dee53026a884e801

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/apimPublisherEventListner.bal (hunk 2 above)
Comment (dewmini) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/apimPublisherEventListner.bal]: Fixed in https://github.com/wso2/carbon-apimgt/pull/4087/commits/e9aae8173ce0104f16c45795dee53026a884e801

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/utils/Utils.bal (hunk 3 above)
Comment (dewmini) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/utils/Utils.bal]: Fixed in https://github.com/wso2/carbon-apimgt/pull/4087/commits/e9aae8173ce0104f16c45795dee53026a884e801

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/utils/Utils.bal (hunk 4 above)
Comment (dewmini) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/utils/Utils.bal]: Fixed in https://github.com/wso2/carbon-apimgt/pull/4087/commits/e9aae8173ce0104f16c45795dee53026a884e801

**Pull request merged**